network/validator/pubsub/validator.go

```diff
@@ -0,0 +1,94 @@
+package validator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+
+	"github.com/onflow/flow-go/network/message"
+	_ "github.com/onflow/flow-go/utils/binstat"
+)
+
+// messagePubKey extracts the public key of the envelope signer from a libp2p message.
+// The location of that key depends on the type of the key, see:
+// https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md
+// This reproduces the exact logic of the private function doing the same decoding in libp2p:
+// https://github.com/libp2p/go-libp2p-pubsub/blob/ba28f8ecfc551d4d916beb748d3384951bce3ed0/sign.go#L77
+func messageSigningID(m *pubsub.Message) (peer.ID, error) {
+	var pubk crypto.PubKey
+
+	// m.From is the original sender of the message (versus `m.ReceivedFrom` which is the last hop which sent us this message)
+	pid, err := peer.IDFromBytes(m.From)
+	if err != nil {
+		return "", err
+	}
+
+	if m.Key == nil {
+		// no attached key, it must be extractable from the source ID
+		pubk, err = pid.ExtractPublicKey()
+		if err != nil {
+			return "", fmt.Errorf("cannot extract signing key: %s", err.Error())
+		}
+		if pubk == nil {
+			return "", fmt.Errorf("cannot extract signing key")
+		}
+	} else {
+		pubk, err = crypto.UnmarshalPublicKey(m.Key)
+		if err != nil {
+			return "", fmt.Errorf("cannot unmarshal signing key: %s", err.Error())
+		}
+
+		// verify that the source ID matches the attached key
+		if !pid.MatchesPublicKey(pubk) {
+			return "", fmt.Errorf("bad signing key; source ID %s doesn't match key", pid)
+		}
+	}
+
+	// the pid either contains or matches the signing pubKey
+	return pid, nil
+}
+
+type MessageValidator func(context.Context, peer.ID, *message.Message) pubsub.ValidationResult
+
+type ValidatorData struct {
+	Message *message.Message
+	From    peer.ID
+}
+
+func TopicValidator(validators ...MessageValidator) pubsub.ValidatorEx {
+	return func(ctx context.Context, receivedFrom peer.ID, rawMsg *pubsub.Message) pubsub.ValidationResult {
+		var msg message.Message
+		// convert the incoming raw message payload to Message type
+		//bs := binstat.EnterTimeVal(binstat.BinNet+":wire>1protobuf2message", int64(len(rawMsg.Data)))
+		err := msg.Unmarshal(rawMsg.Data)
+		//binstat.Leave(bs)
+		if err != nil {
+			return pubsub.ValidationReject
+		}
+
+		from, err := messageSigningID(rawMsg)
+		if err != nil {
+			return pubsub.ValidationReject
+		}
+
+		rawMsg.ValidatorData = ValidatorData{
+			Message: &msg,
+			From:    from,
+		}
+
+		result := pubsub.ValidationAccept
+		for _, validator := range validators {
+			switch res := validator(ctx, from, &msg); res {
```

Comment:
I think we're breaking the contract of a message validator, here: the `from` argument passed to it is supposed to be the direct peer delivering the message, not its origin. Hence, what we should be passing to inner validators here should be `receivedFrom`, not `from`.

It would be good to add a unit test showing delivery of a message through a hop.

Reply:
I intended to pass `from`, not `receivedFrom` (although we can pass both).

Perhaps I need to better explain this PR.

This PR implements a *single* libp2p topic validator, which combines the validation logic of multiple `MessageValidator`s (which are a type specific to Flow codebase). The reason this is necessary is because libp2p only allows registering a single topic validator, yet we may have complex validation logic which should be split into multiple logical validators.

So, with respect to this comment: 

> I think we're breaking the contract of a message validator, here: the from argument passed to it is supposed to be the direct peer delivering the message, not its origin. Hence, what we should be passing to inner validators here should be receivedFrom, not from.

I have defined the `MessageValidator` type, and the contract I have defined along with it is that the argument passed to it is supposed to be the origin of the message. The origin is the more relevant piece of information that most validation logic actually cares about. If a need arises later on, we can also passes `receivedFrom` but currently there is no need for this.

Reply:
> I have defined the MessageValidator type, and the contract I have defined along with it is that the argument passed to it is supposed to be the origin of the message

You've defined the uncommented `MessageValidator` type, which is similar to `pubsub.ValidatorExt`, modulo differences in the message type. The type signature is not enough to differentiate whether the `peerID` argument you're passing is the message's origin or its relayer. It seems natural for the reader to make the same assumptions as in pubsub, where the message's origin is inside the message, and the argument to the validator is the relayer.

In other words: if you want to define `MessageValidator`'s contract, that's fine, but please do so, didactically.

Reply:
Of course, will do